File: ios/Shrank/Database/Migrations.swift
import Foundation
import GRDB

enum AppMigrations {
    static func registerAll(in migrator: inout DatabaseMigrator) {
        migrator.registerMigration("v1-schema") { db in
            // Documents table
            try db.execute(sql: """
                CREATE TABLE documents (
                    id              TEXT PRIMARY KEY,
                    createdAt       TEXT NOT NULL,
                    updatedAt       TEXT NOT NULL,
                    capturedAt      TEXT NOT NULL,
                    syncedAt        TEXT,

                    originalPath    TEXT NOT NULL,
                    thumbnailPath   TEXT NOT NULL,

                    status          TEXT NOT NULL DEFAULT 'pending',
                    processingError TEXT,
                    rawLlmResponse  TEXT,

                    language        TEXT,
                    sender          TEXT,
                    senderNormalized TEXT,
                    documentDate    TEXT,
                    documentType    TEXT,
                    subject         TEXT,
                    extractedText   TEXT,

                    amounts         TEXT,
                    dates           TEXT,
                    referenceIds    TEXT,
                    tags            TEXT,

                    confidence      REAL
                )
                """)

            // FTS5 full-text search index
            try db.execute(sql: """
                CREATE VIRTUAL TABLE documents_fts USING fts5(
                    sender,
                    subject,
                    extractedText,
                    tags,
                    content='documents',
                    content_rowid='rowid',
                    tokenize='unicode61 remove_diacritics 2'
                )
                """)

            // FTS5 sync triggers
            try db.execute(sql: """
                CREATE TRIGGER documents_ai AFTER INSERT ON documents BEGIN
                    INSERT INTO documents_fts(rowid, sender, subject, extractedText, tags)
                    VALUES (new.rowid, new.sender, new.subject, new.extractedText, new.tags);
                END
                """)

            try db.execute(sql: """
                CREATE TRIGGER documents_ad AFTER DELETE ON documents BEGIN
                    INSERT INTO documents_fts(documents_fts, rowid, sender, subject, extractedText, tags)
                    VALUES ('delete', old.rowid, old.sender, old.subject, old.extractedText, old.tags);
                END
                """)

            try db.execute(sql: """
                CREATE TRIGGER documents_au AFTER UPDATE ON documents BEGIN
                    INSERT INTO documents_fts(documents_fts, rowid, sender, subject, extractedText, tags)
                    VALUES ('delete', old.rowid, old.sender, old.subject, old.extractedText, old.tags);
                    INSERT INTO documents_fts(rowid, sender, subject, extractedText, tags)
                    VALUES (new.rowid, new.sender, new.subject, new.extractedText, new.tags);
                END
                """)

            // Entities table
            try db.execute(sql: """
                CREATE TABLE entities (
                    id              TEXT PRIMARY KEY,
                    entityType      TEXT NOT NULL,
                    value           TEXT NOT NULL,
                    displayName     TEXT,
                    metadata        TEXT,
                    createdAt       TEXT NOT NULL,
                    UNIQUE(entityType, value)
                )
                """)

            // Document-entity links
            try db.execute(sql: """
                CREATE TABLE document_entities (
                    documentId      TEXT NOT NULL REFERENCES documents(id) ON DELETE CASCADE,
                    entityId        TEXT NOT NULL REFERENCES entities(id) ON DELETE CASCADE,
                    role            TEXT NOT NULL,
                    confidence      REAL DEFAULT 1.0,
                    PRIMARY KEY (documentId, entityId, role)
                )
                """)

            // Document-document edges
            try db.execute(sql: """
                CREATE TABLE document_edges (
                    sourceId        TEXT NOT NULL REFERENCES documents(id) ON DELETE CASCADE,
                    targetId        TEXT NOT NULL REFERENCES documents(id) ON DELETE CASCADE,
                    relationType    TEXT NOT NULL,
                    confidence      REAL DEFAULT 1.0,
                    inferredBy      TEXT NOT NULL,
                    createdAt       TEXT NOT NULL,
                    PRIMARY KEY (sourceId, targetId, relationType)
                )
                """)

            // Graph indexes
            try db.execute(sql: "CREATE INDEX idx_doc_entities_entity ON document_entities(entityId)")
            try db.execute(sql: "CREATE INDEX idx_doc_entities_doc ON document_entities(documentId)")
            try db.execute(sql: "CREATE INDEX idx_doc_edges_source ON document_edges(sourceId)")
            try db.execute(sql: "CREATE INDEX idx_doc_edges_target ON document_edges(targetId)")

            // Upload queue
            try db.execute(sql: """
                CREATE TABLE upload_queue (
                    id              TEXT PRIMARY KEY,
                    localImagePath  TEXT NOT NULL,
                    capturedAt      TEXT NOT NULL,
                    status          TEXT NOT NULL DEFAULT 'pending',
                    retryCount      INTEGER DEFAULT 0,
                    lastAttemptAt   TEXT,
                    errorMessage    TEXT
                )
                """)

            // Sync state (single-row table)
            try db.execute(sql: """
                CREATE TABLE sync_state (
                    deviceId        TEXT PRIMARY KEY,
                    lastSyncCursor  TEXT,
                    lastSyncAt      TEXT
                )
                """)

            // Document indexes for common queries
            try db.execute(sql: "CREATE INDEX idx_documents_captured ON documents(capturedAt)")
            try db.execute(sql: "CREATE INDEX idx_documents_status ON documents(status)")
            try db.execute(sql: "CREATE INDEX idx_documents_sender ON documents(senderNormalized)")
            try db.execute(sql: "CREATE INDEX idx_upload_queue_status ON upload_queue(status)")
        }
    }
}

File: server/src/db/migrations.rs
use rusqlite::Connection;

pub fn run(conn: &Connection) -> Result<(), rusqlite::Error> {
    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS _migrations (
            version INTEGER PRIMARY KEY,
            applied_at TEXT NOT NULL DEFAULT (datetime('now'))
        );",
    )?;

    let current: i64 = conn
        .query_row(
            "SELECT COALESCE(MAX(version), 0) FROM _migrations",
            [],
            |row| row.get(0),
        )?;

    if current < 1 {
        migrate_v1(conn)?;
        conn.execute("INSERT INTO _migrations (version) VALUES (1)", [])?;
        tracing::info!("applied migration v1: initial schema");
    }

    if current < 2 {
        migrate_v2(conn)?;
        conn.execute("INSERT INTO _migrations (version) VALUES (2)", [])?;
        tracing::info!("applied migration v2: add ocr_markdown column");
    }

    Ok(())
}

fn migrate_v1(conn: &Connection) -> Result<(), rusqlite::Error> {
    conn.execute_batch(
        "
        -- Documents: the primary entity
        CREATE TABLE documents (
            id              TEXT PRIMARY KEY,
            created_at      TEXT NOT NULL,
            updated_at      TEXT NOT NULL,
            captured_at     TEXT NOT NULL,
            synced_at       TEXT,

            original_path   TEXT NOT NULL,
            thumbnail_path  TEXT NOT NULL,

            status          TEXT NOT NULL DEFAULT 'pending',
            processing_error TEXT,
            raw_llm_response TEXT,

            language        TEXT,
            sender          TEXT,
            sender_normalized TEXT,
            document_date   TEXT,
            document_type   TEXT,
            subject         TEXT,
            extracted_text  TEXT,

            amounts         TEXT,
            dates           TEXT,
            reference_ids   TEXT,

            tags            TEXT,
            confidence      REAL
        );

        -- Full-text search index
        CREATE VIRTUAL TABLE documents_fts USING fts5(
            sender,
            subject,
            extracted_text,
            tags,
            content='documents',
            content_rowid='rowid',
            tokenize='unicode61 remove_diacritics 2'
        );

        -- Triggers to keep FTS in sync
        CREATE TRIGGER documents_ai AFTER INSERT ON documents BEGIN
            INSERT INTO documents_fts(rowid, sender, subject, extracted_text, tags)
            VALUES (new.rowid, new.sender, new.subject, new.extracted_text, new.tags);
        END;

        CREATE TRIGGER documents_ad AFTER DELETE ON documents BEGIN
            INSERT INTO documents_fts(documents_fts, rowid, sender, subject, extracted_text, tags)
            VALUES ('delete', old.rowid, old.sender, old.subject, old.extracted_text, old.tags);
        END;

        CREATE TRIGGER documents_au AFTER UPDATE ON documents BEGIN
            INSERT INTO documents_fts(documents_fts, rowid, sender, subject, extracted_text, tags)
            VALUES ('delete', old.rowid, old.sender, old.subject, old.extracted_text, old.tags);
            INSERT INTO documents_fts(rowid, sender, subject, extracted_text, tags)
            VALUES (new.rowid, new.sender, new.subject, new.extracted_text, new.tags);
        END;

        -- Entities: things that appear across multiple documents
        CREATE TABLE entities (
            id              TEXT PRIMARY KEY,
            entity_type     TEXT NOT NULL,
            value           TEXT NOT NULL,
            display_name    TEXT,
            metadata        TEXT,
            created_at      TEXT NOT NULL,
            UNIQUE(entity_type, value)
        );

        -- Links between documents and entities
        CREATE TABLE document_entities (
            document_id     TEXT NOT NULL REFERENCES documents(id) ON DELETE CASCADE,
            entity_id       TEXT NOT NULL REFERENCES entities(id) ON DELETE CASCADE,
            role            TEXT NOT NULL,
            confidence      REAL DEFAULT 1.0,
            PRIMARY KEY (document_id, entity_id, role)
        );

        -- Direct document-to-document relationships
        CREATE TABLE document_edges (
            source_id       TEXT NOT NULL REFERENCES documents(id) ON DELETE CASCADE,
            target_id       TEXT NOT NULL REFERENCES documents(id) ON DELETE CASCADE,
            relation_type   TEXT NOT NULL,
            confidence      REAL DEFAULT 1.0,
            inferred_by     TEXT NOT NULL,
            created_at      TEXT NOT NULL,
            PRIMARY KEY (source_id, target_id, relation_type)
        );

        -- Indexes for graph traversal
        CREATE INDEX idx_doc_entities_entity ON document_entities(entity_id);
        CREATE INDEX idx_doc_entities_doc ON document_entities(document_id);
        CREATE INDEX idx_doc_edges_source ON document_edges(source_id);
        CREATE INDEX idx_doc_edges_target ON document_edges(target_id);

        -- Sync state
        CREATE TABLE sync_cursors (
            client_id       TEXT PRIMARY KEY,
            last_sync_at    TEXT NOT NULL,
            last_document_id TEXT
        );

        -- Additional indexes for common queries
        CREATE INDEX idx_documents_status ON documents(status);
        CREATE INDEX idx_documents_sender ON documents(sender_normalized);
        CREATE INDEX idx_documents_type ON documents(document_type);
        CREATE INDEX idx_documents_date ON documents(document_date);
        CREATE INDEX idx_documents_updated ON documents(updated_at);
        ",
    )?;

    Ok(())
}

fn migrate_v2(conn: &Connection) -> Result<(), rusqlite::Error> {
    conn.execute_batch(
        "ALTER TABLE documents ADD COLUMN ocr_markdown TEXT;",
    )?;
    Ok(())
}
